src/coprocessor/dag/expr/arithmetic.rs

```diff
@@ -0,0 +1,401 @@
+        let result: Result<Decimal> = (lhs.as_ref() + rhs.as_ref()).into();
+        result.map(Cow::Owned).map(Some)
+    }
+
+    pub fn plus_int(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<i64>> {
+        let lhs = try_opt!(self.children[0].eval_int(ctx, row));
+        let rhs = try_opt!(self.children[1].eval_int(ctx, row));
+        let lus = mysql::has_unsigned_flag(self.children[0].get_tp().get_flag());
+        let rus = mysql::has_unsigned_flag(self.children[1].get_tp().get_flag());
+        let res = match (lus, rus) {
+            (true, true) => try!((lhs as u64).checked_add(rhs as u64).ok_or(Error::Overflow)),
+            (true, false) => if rhs >= 0 {
+                try!((lhs as u64).checked_add(rhs as u64).ok_or(Error::Overflow))
+            } else {
+                let rhs = try!(rhs.checked_neg().ok_or(Error::Overflow)) as u64;
+                try!((lhs as u64).checked_sub(rhs).ok_or(Error::Overflow))
+            },
+            (false, true) => if lhs >= 0 {
+                try!((lhs as u64).checked_add(rhs as u64).ok_or(Error::Overflow))
+            } else {
+                let lhs = try!(lhs.checked_neg().ok_or(Error::Overflow)) as u64;
+                try!((rhs as u64).checked_sub(lhs).ok_or(Error::Overflow))
+            },
+            (false, false) => try!(lhs.checked_add(rhs).ok_or(Error::Overflow)) as u64,
+        };
+        Ok(Some(res as i64))
+    }
+
+    pub fn minus_real(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<f64>> {
+        let lhs = try_opt!(self.children[0].eval_real(ctx, row));
+        let rhs = try_opt!(self.children[1].eval_real(ctx, row));
+        let res = lhs - rhs;
+        if !res.is_finite() {
+            return Err(Error::Overflow);
+        }
+        Ok(Some(res))
+    }
+
+    pub fn minus_decimal<'a, 'b: 'a>(
+        &'b self,
+        ctx: &StatementContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Decimal>>> {
+        let lhs = try_opt!(self.children[0].eval_decimal(ctx, row));
+        let rhs = try_opt!(self.children[1].eval_decimal(ctx, row));
+        let result: Result<Decimal> = (lhs.as_ref() - rhs.as_ref()).into();
+        result.map(Cow::Owned).map(Some)
+    }
+
+    pub fn minus_int(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<i64>> {
+        let lhs = try_opt!(self.children[0].eval_int(ctx, row));
+        let rhs = try_opt!(self.children[1].eval_int(ctx, row));
+        let lus = mysql::has_unsigned_flag(self.children[0].get_tp().get_flag());
+        let rus = mysql::has_unsigned_flag(self.children[1].get_tp().get_flag());
+        let res = match (lus, rus) {
+            (true, true) => try!((lhs as u64).checked_sub(rhs as u64).ok_or(Error::Overflow)),
+            (true, false) => if rhs >= 0 {
+                try!((lhs as u64).checked_sub(rhs as u64).ok_or(Error::Overflow))
+            } else {
+                let rhs = try!(rhs.checked_neg().ok_or(Error::Overflow)) as u64;
```

Comment:
You can use `rhs.checked_neg().and_then(|rhs| (rhs as u64).checked_add(lhs as u64))`. And change L109 to `res.ok_or(Error::Overflow).map(Some)`.